lib/job-iteration/iteration.rb

```diff
@@ -288,6 +261,44 @@ def handle_completed(completed)
       raise "Unexpected thrown value: #{completed.inspect}"
     end
 
+    def cursor_position_from_job_data(job_data)
+      # Backwards compatibility for jobs interrupted before cursor serialization feature shipped
+      return job_data["cursor_position"] unless job_data.key?("serialized_cursor_position")
+
+      deserialize_cursor(job_data.fetch("serialized_cursor_position"))
+    end
+
+    def serialize_cursor(cursor)
+      ActiveJob::Arguments.serialize([cursor]).first
+    rescue ActiveJob::SerializationError
+      raise if job_iteration_enforce_serializable_cursors
+
+      Deprecation.warn(<<~DEPRECATION_MESSAGE)
+        The Enumerator returned by #{self.class.name}#build_enumerator yielded a cursor which is unsafe to serialize.
+
+        TBD MENTION SERIALIZERS!
+
+        This will raise starting in version #{Deprecation.deprecation_horizon} of #{Deprecation.gem_name}!"
+      DEPRECATION_MESSAGE
+
+      cursor
+    end
+    alias_method :assert_valid_cursor!, :serialize_cursor
+
+    def deserialize_cursor(cursor)
+      ActiveJob::Arguments.deserialize([cursor]).first
+    rescue ActiveJob::SerializationError
+      raise if job_iteration_enforce_serializable_cursors
+
+      Deprecation.warn(<<~DEPRECATION_MESSAGE)
+        The cursor stored for #{self.class.name} could not be deserialized properly. Falling back to legacy behavior.
+
+        This will raise starting in version #{Deprecation.deprecation_horizon} of #{Deprecation.gem_name}!"
+      DEPRECATION_MESSAGE
+
+      cursor
```

Comment:
I should add a test for this somehow.

Reply:
I realized we should never store an unserialized cursor as `serialized_cursor_position`. I'm refactoring so that we only store `serialized_cursor_position` if it succeeded, and only deserialize it if it is there (falling back to unserialized `cursor_position` otherwise).